core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -347,13 +338,11 @@ class BrokerToControllerRequestThread(
         response.versionMismatch)
       queueItem.callback.onComplete(response)
     } else if (response.wasDisconnected()) {
-      updateControllerAddress(null)
       requestQueue.putFirst(queueItem)
     } else if (response.responseBody().errorCounts().containsKey(Errors.NOT_CONTROLLER)) {
```

Comment:
Can we detect that the request has been sending to an obsolete controller through the response error code, such that we could call "updateControllerAddress(null)" similar here to trigger getting an updated controller in doWork()?

Reply:
Even if we do that, we'd still need to call the controllerNodeProvider for getting the up-to-date controller node.
Compared with only querying the controllerNodeProvider upon failures, I feel that always querying it is a cleaner design.

Reply:
My original thought was that if we could detect the obsolete controller from the response, we could just updateControllerAddress as necessary as the original logic and no need to call controllerNodeProvider every time generating a new request. But as you said in the description that  "controllerNodeProvider.get()" is not an expensive operation, it should be fine either way. 

Another thing is that if the request is failed due to incorrect controller, we could queue the request again (requestQueue.putFirst(queueItem)) as a retriable error like Errors.NOT_CONTROLLER instead of waiting for other retry logics.

Reply:
Re-queuing the items may cause the request to go out-of-order, which can have unexpected consequences.
So I won't go that route.